### ISSUE TITLE
Soil Disturbance: should allow selection of crop x bed affected

### ISSUE THREAD

**Issue by braughtg** [FarmData2/FarmData2#372]
The Soil Disturbance form currently has a checkbox for "Termination Event," which appears only if the location/beds selected contain active plant assets.  When this box is checked all plants in the location/beds are archived.

In practice an area or bed can contain many different plant assets and it is a common occurrence to terminate just one or a few of those plants, but not all of them.  In addition, a plant may be in multiple beds, and only one or a few, but not all of the beds could be terminated.

Thus, the "Termination Event" indication must be done on a plant x bed basis.  To do this, the "Termination Event" checkbox should be expanded into a `PicklistBase` element that displays the list of plants in the selected location/beds.  There should be one entry for each plant x bed (i.e. if a plant is in two beds, then it appears twice in the list).  Each row should display the crop name, the bed name and the date it was planted in the ground (direct seeded or transplanted).  If a location does not have beds associated with it then the plant will be listed only once. Like the "Termination Event" checkbox, this `PicklistBase` should only appear if there are plants in the selected location/beds.

Currently the "Termination Event" checkbox is only displayed if a call to `farmosUtil.getPlantAssets` returns some records.  Those records contain sufficient information to populate the plant x beds list.  All rows of of this list should be selected by default.

If a plant x bed row is selected in a submission, the bed should be removed from the plant's locations.  If the plant has no remaining locations, then the plant should be archived.

The e2e test for the "Soil Disturbance" entry point will need to be updated.

**Comment by Shahir-47** [comment]
Hi Professor Braught,

I’ve been working through some design options and wanted to get your input on which approach you think would be most suitable. Below are the four designs I’ve come up with:

1.  
![Design 1](https://github.com/user-attachments/assets/15f910c4-44ab-4c50-89cf-d754a8799246)

2.  
![Design 2](https://github.com/user-attachments/assets/3be442c2-0f4d-4007-9031-a9d0d1461f9c)

3.  
![Design 3](https://github.com/user-attachments/assets/6c10a3d4-1c43-40f4-a86d-1d0745036819)

4.  
![Design 4](https://github.com/user-attachments/assets/f6ddba65-23ac-490f-923a-c32e05e28dc0)

Could you please advise which design you prefer, or if you have any additional feedback?

Thank you!

**Comment by braughtg** [comment]
> Could you please advise which design you prefer, or if you have any additional feedback?

I like #3 because I think the centered title is most consistent with our other forms and the vertical design will work better than the horizontal design in 2 and 4 on mobile.

I think we should use "Crop" instead of "Plant" to be consistent with what we call it when we are doing the seeding or transplanting.

**Comment by Shahir-47** [comment]
@braughtg I have a few follow-up questions regarding plant assets and their bed locations:

- To identify if a plant has no remaining locations, would you check for the `location` attribute in the plant asset object to be `null` or the `beds` attribute to be non-existent, an empty array, or `null`?
- If a plant asset has no beds, should it be displayed? If it should, should I set the `bed` attribute in `form.affectedPlants` to `null`, and would that value be displayed in the "Beds" column of the `PicklistBase` table for that particular plant with no beds?
- If this plant asset with no beds is selected, should I directly archive it?
- For plant assets that do have beds, if I remove their last bed, should the plant asset have the `bed` attribute set to `null`/removed, or should I archive the plant asset in this case as well?

Let me know what you think!

**Comment by braughtg** [comment]
> To identify if a plant has no remaining locations, would you check for the `location` attribute in the plant asset object to be `null` or the `beds` attribute to be non-existent, an empty array, or `null`?
>

Any plant asset with no remaining locations should have already been archived.  Thus, the API should not return any plant assets that have no remaining locations.

> If a plant asset has no beds, should it be displayed? If it should, should I set the `bed` attribute in `form.affectedPlants` to `null`, and would that value be displayed in the "Beds" column of the `PicklistBase` table for that particular plant with no beds?
> 

If a plant asset exists in the selected location then it should be listed in the table.

If there are plant assets in the location, some with beds and some without, then we have two choices for when the ones without beds should be displayed.  the two choices are always no matter which beds are picked or only when no beds are picked.  I'm leaning toward always with the use of `N/A` as the bed for the plant assets without beds specified.  Those with beds specified will list the bed.

If none of the plant assets in the location have beds, then I think we should omit the bed column from the `PickList` when it is displayed.

> If this plant asset with no beds is selected, should I directly archive it?
>

Yes, if a plant asset with no beds is checked as a termination event, then the plant asset should be archived by the transplanting operation.
 
> For plant assets that do have beds, if I remove their last bed, should the plant asset have the `bed` attribute set to `null`/removed, or should I archive the plant asset in this case as well?
> 

If a plant asset has beds associated with it and all beds are checked as termination events, then the plant asset should be archived by the transplanting operation.

**Comment by Shahir-47** [comment]
Hi @braughtg,

I've been working on the soil disturbance feature, and I've reached a point where it would be beneficial to have a function that updates plant assets, specifically `farmosUtil.updatePlantAsset()`. 

In my draft PR [here](https://github.com/FarmData2/FarmData2/pull/388), I've made modifications to `lib.js` that rely on updating plant assets, particularly when handling bed removals during termination events. As of now, this functionality doesn't exist, so I wanted to propose adding `updatePlantAsset()` to `farmosUtil.js`.

Could you take a look at my proposed changes when you get the chance? I'd really appreciate your opinion on whether adding `updatePlantAsset()` makes sense, or if there's a better approach you would suggest.

Thanks in advance for your time!

Best,
Shahir

**Comment by braughtg** [comment]
I think that having an `updatePlantAsset` function in `farmosUtil.js` makes good sense.  But let's split it off into its own PR.  That PR should just add the `updatePlantAsset` function and the associated tests for it.  We can review and merge it as its own PR.  Then you can use it in #388.

Please create an issue for creating and testing `updatePlantAsset`, add it to the project board and reference it in the new PR that you create for `updatePlantAsset`.

**Comment by Shahir-47** [comment]
@braughtg,

I'm working on implementing termination events in the Soil Disturbance form. We need to create logs for terminating plants on a plant x bed basis, where we remove beds from plant assets and archive them if they have no beds left.

Currently, none of the existing log creation functions in `farmosUtil.js` fit our needs:

- `createSeedingLog` is for seeding activities.
- `createSoilDisturbanceActivityLog` is tailored toward equipment used to disrupt the soil.
- `createWinterKillActivityLog` is specific to winter kill events with cover crops.
- `createTransplantingActivityLog` is for transplanting trays/plants.

Given this, I think we need to add a new function to `farmosUtil.js` for termination events. It would be similar to `createWinterKillActivityLog` but for general termination. Here's a proposed function:
```js
/**
 * Create a new activity log (`log--activity`) for a termination event.
 *
 * @param {string} terminationDate - The date the termination took place.
 * @param {string} locationName - The location where the termination took place.
 * @param {Array<string>} bedNames - The names of the bed(s) where the termination occurred.
 * @param {Object} plantAsset - The plant asset being terminated.
 * @returns {Object} The new activity log.
 */
export async function createTerminationActivityLog(
  terminationDate,
  locationName,
  bedNames = [],
  plantAsset
) {
  const locationsArray = await getPlantingLocationObjects([
    locationName,
    ...bedNames,
  ]);
  const logCategoriesArray = await getLogCategoryObjects(['termination']);

  const cropIdToTermMap = await getCropIdToTermMap();
  const logName =
    dayjs(terminationDate).format('YYYY-MM-DD') +
    '_tr_' +
    plantAsset.relationships.plant_type
      .map((crop) => cropIdToTermMap.get(crop.id).attributes.name)
      .join('_');

  const activityLogData = {
    type: 'log--activity',
    attributes: {
      name: logName,
      timestamp: dayjs(terminationDate).format(),
      status: 'done',
      is_movement: true,
      purchase_date: dayjs(terminationDate).format(),
    },
    relationships: {
      location: locationsArray,
      asset: [{ type: 'asset--plant', id: plantAsset.id }],
      category: logCategoriesArray,
    },
  };

  const farm = await getFarmOSInstance();
  const activityLog = farm.log.create(activityLogData);
  await farm.log.send(activityLog);

  return activityLog;
}
````
This function creates an activity log for termination events, allowing us to handle termination as required.

What do you think about adding this function to `farmosUtil.js`?

**Comment by Shahir-47** [comment]
@braughtg,

If adding a new function isn’t the best option, we can also use the existing `createSoilDisturbanceActivityLog` function to handle termination events. Here’s an example of how it could work:

```js
import * as farmosUtil from '@libs/farmosUtil/farmosUtil';

// Helper function to compare two sets
function setsAreEqual(a, b) {
  if (a.size !== b.size) return false;
  for (let item of a) {
    if (!b.has(item)) return false;
  }
  return true;
}

/**
 * Create the farmOS records (asset, quantities and log) to represent
 * a soil disturbance
 *
 * @param {Object} form the form containing the data from the entry point.
 * @returns {Promise} a promise that resolves when the records are successfully created.
 * The returned value is an object containing the asset, quantities and log that
 * were sent to the server.  This object has the properties shown below
 * where i indicates passes:
 * ```Javascript
 * {
 *   equipment: [ {asset--equipment} ],
 *   affectedPlants: [ {asset--plant} ],
 *   depth(i): {quantity--standard},
 *   speed(i): {quantity--standard},
 *   area(i): {quantity--standard},
 *   activityLog(i): {log--activity},
 * }
 * ```
 * @throws {Error} if an error occurs while creating the farmOS records.
 */
async function submitForm(formData) {
  try {
    let ops = [];
    const equipmentAssets = [];
    const plantAssets = [];

    // group by plant id
    const uniquePlantAssetIds = new Set(
      Array.from(formData.picked.values()).map((item) => item.row.plantAssetId)
    );

    // Operation to archive plant assets based on your criteria
    const archivePlantsOp = {
      name: 'archivePlants',
      do: async () => {
        for (let plantAssetId of uniquePlantAssetIds) {
          try {
            // Fetch the plant asset
            const plantAsset = await farmosUtil.getPlantAsset(plantAssetId);
            const plantBeds = new Set(plantAsset.attributes.beds || []);

            // Get selected beds for this plant asset
            const selectedBeds = new Set(
              Array.from(formData.picked.values())
                .filter((item) => item.row.plantAssetId === plantAssetId)
                .map((item) => item.row.bed)
            );

            // Check if the plant asset has no beds
            if (plantBeds.size === 0) {
              // Plant asset has no beds, archive it
              const archivedPlantAsset = await farmosUtil.archivePlantAsset(
                plantAssetId,
                true
              );
              plantAssets.push(archivedPlantAsset);
            } else if (setsAreEqual(plantBeds, selectedBeds)) {
              // All beds are selected, archive the plant asset
              const archivedPlantAsset = await farmosUtil.archivePlantAsset(
                plantAssetId,
                true
              );
              plantAssets.push(archivedPlantAsset);
            } else {
              // Not all beds are selected, do not archive the plant asset
              // Add plant asset to plantAssets array for use in the termination log
              plantAssets.push(plantAsset);
            }
          } catch (error) {
            console.error(
              `Error processing plant asset with id ${plantAssetId}:`,
              error
            );
            throw error;
          }
        }
      },
      undo: async () => {
        for (let plantAsset of plantAssets) {
          try {
            // Unarchive the plant asset
            await farmosUtil.archivePlantAsset(plantAsset.id, false);
          } catch (error) {
            console.error(
              `Error unarchiving plant asset with id ${plantAsset.id}:`,
              error
            );
          }
        }
      },
    };
    ops.push(archivePlantsOp);

    const terminationLog = {
      name: 'terminationLog',
      do: async () => {
        // Extract bed names from the picked items
        const bedNames = Array.from(formData.picked.values()).map(
          (item) => item.row.bed
        );

        return await farmosUtil.createSoilDisturbanceActivityLog(
          formData.date,
          formData.location,
          bedNames,
          formData.picked.size > 0 ? ['tillage', 'termination'] : ['tillage'],
          plantAssets,
          [],
          [],
          'Terminate selected plants in beds'
        );
      },
      undo: async (results) => {
        await farmosUtil.deleteSoilDisturbanceActivityLog(
          results['terminationLog'].id
        );
      },
    };
    ops.push(terminationLog);

    const equipmentMap = await farmosUtil.getEquipmentNameToAssetMap();
    for (const equipmentName of formData.equipment) {
      equipmentAssets.push(equipmentMap.get(equipmentName));
    }

    for (let i = 0; i < formData.passes; i++) {
      const depthQuantity = {
        name: 'depthQuantity' + i,
        do: async () => {
          return await farmosUtil.createStandardQuantity(
            'length',
            formData.depth,
            'Depth',
            'INCHES'
          );
        },
        undo: async (results) => {
          await farmosUtil.deleteStandardQuantity(
            results['depthQuantity' + i].id
          );
        },
      };
      ops.push(depthQuantity);

      const speedQuantity = {
        name: 'speedQuantity' + i,
        do: async () => {
          return await farmosUtil.createStandardQuantity(
            'rate',
            formData.speed,
            'Speed',
            'MPH'
          );
        },
        undo: async (results) => {
          await farmosUtil.deleteStandardQuantity(
            results['speedQuantity' + i].id
          );
        },
      };
      ops.push(speedQuantity);

      const areaQuantity = {
        name: 'areaQuantity' + i,
        do: async () => {
          return await farmosUtil.createStandardQuantity(
            'ratio',
            formData.area,
            'Area',
            'PERCENT'
          );
        },
        undo: async (results) => {
          await farmosUtil.deleteStandardQuantity(
            results['areaQuantity' + i].id
          );
        },
      };
      ops.push(areaQuantity);

      const activityLog = {
        name: 'activityLog' + i,
        do: async (results) => {
          return await farmosUtil.createSoilDisturbanceActivityLog(
            formData.date,
            formData.location,
            formData.beds,
            formData.picked.size > 0 ? ['tillage', 'termination'] : ['tillage'],
            plantAssets,
            [
              results['depthQuantity' + i],
              results['speedQuantity' + i],
              results['areaQuantity' + i],
            ],
            equipmentAssets,
            'Pass ' +
              (i + 1) +
              ' of ' +
              formData.passes +
              '. ' +
              formData.comment
          );
        },
        undo: async (results) => {
          await farmosUtil.deleteSoilDisturbanceActivityLog(
            results['activityLog' + i].id
          );
        },
      };
      ops.push(activityLog);
    }

    const result = await farmosUtil.runTransaction(ops);
    result['equipment'] = equipmentAssets;

    return result;
  } catch (error) {
    console.error('SoilDisturbance: \n' + error.message);
    console.error(error);

    let errorMsg = 'Error creating Soil Disturbance records.';

    for (const key of Object.keys(error.results)) {
      if (error.results[key]) {
        errorMsg +=
          '\n  Result of operation ' + key + ' could not be cleaned up.';
        if (
          error.results[key].attributes &&
          error.results[key].attributes.name
        ) {
          errorMsg += '\n   Manually delete log or asset with:';
          errorMsg += '\n     name: ' + error.results[key].attributes.name;
        } else {
          errorMsg += '\n   May be safely ignored';
        }
      }
    }

    throw Error(errorMsg, error);
  }
}

export const lib = {
  submitForm,
};
```
This version works with `createSoilDisturbanceActivityLog` to handle the termination events and archives plants if necessary based on bed selection. I haven’t written tests yet, so I’m not 100% sure how well it works in practice.

Let me know if you think this works or if we should go with adding a new function!

**Comment by braughtg** [comment]
> This function creates an activity log for termination events, allowing us to handle termination as required.
> 
> What do you think about adding this function to `farmosUtil.js`?

What do you think about just using the `log--activity` that is being created to represent the soil disturbance? That log could assign the new location array (omitting the terminated beds) and could include `termination` along with `tillage` as the log categories?

What advantage/disadvantage would there be to this as compared to creating a separate `log--activity` for the termination?

**Comment by Shahir-47** [comment]
> ```js
>is_movement: true,
> ```

Yes, we can use the `createSoilDisturbanceActivityLog` for the termination event. However, the `is_movement` attribute isn’t set to `true` in the log created by `createSoilDisturbanceActivityLog`. 

We would need to adjust that if we want to reflect the movement (or termination) of plants from specific beds.

Let me know what you think.

**Comment by braughtg** [comment]
> Yes, we can use the `createSoilDisturbanceActivityLog` for the termination event. However, the `is_movement` attribute isn’t set to `true` in the log created by `createSoilDisturbanceActivityLog`.

When `createSoilDisturbanceActivityLog` is called from `lib.js` in the `soil_disturbance` entry point, it is passed an array of log categories (line 123 in current `development` branch).  These log categories will always contain `tillage` and currently if the planting is terminated (e.g. one bed) it will contain `termination` also.  So I think we can adapt this approach.

We'll have 3 cases for each plant asset affected by the soil disturbance:
- No beds associated with the plant-asset are terminated: In this case, the plant asset is not archived and we just get the soil disturbance log, which is not a movement.
- Some but not all beds associated with the plant asset are terminated: In this case, the plant asset is not terminated, but the soil disturbance log is a movement that removes the terminated beds from the plant asset's locations.  The log comment should be appended to indicate the location and beds that were terminated (e.g. `This disturbance terminated the crop in beds: ALF-1, B-3, ...`).
- All beds, or the location (if beds are not indicated), are terminated: In this case the plant asset is archived and the soil disturbance log is a movement with no locations (i.e. the plant no longer exists anywhere).  The log comment should be appended to indicate the beds or location that were terminated (e.g. `This disturbance terminated the crop in beds: ALF-1, B-3, ...`).

**Comment by braughtg** [comment]
Additional discussion of this issue has raised a design issue that needs input from users.  I'll connect with Matt from the farm and see what he thinks.

The issue at hand is that there seem to be two different scopes of what the soil disturbance might apply to.
- If we select a location and possibly some beds, but no termination events for specific crop x bed combinations, then the disturbance would seem to apply to all of the plants in the selected areas. (e.g. weeding paths, spraying, etc).
- But if we select a location and possibly some beds and check some specific crop x bed combinations, then it would seem that the disturbance applies to all plants in the selected areas but only some of them are terminated.  It is this case that seems confusing.
  - It is not clear that this situation would ever arise.  If a piece of equipment is setup to do termination (e.g. plowing under a crop) then the disturbance would likely only apply to the crops x beds that are selected for termination and not to any of the other plants in the selected location and beds.
  - So it seems to make sense that such a disturbance would only apply to the plants that are being terminated.  But this would mean that the form behaves substantially differently in these two situations which does not seem intuitive and may be confusing.

**Comment by braughtg** [comment]
> The issue at hand is that there seem to be two different scopes of what the soil disturbance might apply to.

The following is a proposal for how to resolve the issue of distinguishing these scopes.  

If `farmosUtil.getPlantAssets` returns some records (i.e. there are plant assets in the selected location) then:
- The crop x bed picklist (as described in earlier comments) will be displayed. 
- In addition, a single "Termination Event" check box will be displayed.

The crop x bed picklist is used to choose the crop x bed combinations to which the soil disturbance applies.

The "Termination Event" check box will be used to indicate if the soil disturbance is a termination event.  
- If the termination event box is checked, then all of the crop x bed combinations that are selected will be associated with a disturbance log, and they will be terminated.
- If the termination event box is not checked, then all of the crop x bed combinations that are selected will be associated with a disturbance log, but they will not be terminated.

Here are some mockups of what this could look like:

When there are active plants in the selected location the crop x bed picklist and the termination event checkbox appear:
![Image](https://github.com/user-attachments/assets/34da3678-5b57-455a-b416-a95c0d6f2f31)

When there are not active plants in the selected location just the bed picker appears (if there are beds defined in the selected location):
![Image](https://github.com/user-attachments/assets/04ef4331-18c4-4c80-b00d-84045fa3d217)

Some default behavior that should be considered:
- If there are no active plantings, all beds if any, should be selected by default.
- If there are active plantings no rows in the crop x bed picklist should be selected by default.
